gcloud/search/document.py

```diff
@@ -0,0 +1,351 @@
+    See:
+    https://cloud.google.com/search/reference/rest/google/cloudsearch/v1/FieldValue
+
+    :type geo_value: tuple, (float, float)
+    :param geo_value: latitude, longitude
+    """
+
+    value_type = 'geo'
+
+    def __init__(self, geo_value):
+        self.geo_value = geo_value
+
+
+class Field(object):
+    """Fields hold values for a given document
+
+    See:
+    https://cloud.google.com/search/reference/rest/google/cloudsearch/v1/FieldValueList
+
+    :type name: string
+    :param name: field name
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.values = []
+
+    def add_value(self, value, **kw):
+        """Add a value to the field.
+
+        Selects type of value instance based on type of ``value``.
+
+        :type value: string, integer, float, datetime, or tuple (float, float)
+        :param value: the field value to add.
+
+        :param kw: extra keyword arguments to be passed to the value instance
+                   constructor.
+
+        :raises: ValueError if unable to match the type of ``value``.
+        """
+        if isinstance(value, six.string_types):
+            self.values.append(StringValue(value, **kw))
+        elif isinstance(value, (six.integer_types, float)):
+            self.values.append(NumberValue(value, **kw))
+        elif isinstance(value, datetime.datetime):
+            self.values.append(TimestampValue(value, **kw))
+        elif isinstance(value, tuple):
+            self.values.append(GeoValue(value, **kw))
+        else:
+            raise ValueError("Couldn't determine value type: %s" % (value,))
+
+
+class Document(object):
+    """Documents hold values for search within indexes.
+
+    See:
+    https://cloud.google.com/search/reference/rest/v1/projects/indexes/documents
+
+    :type name: string
+    :param name: the name of the document
+
+    :type index: :class:`gcloud.search.index.Index`
+    :param index: the index to which the document belongs.
+
+    :type rank: positive integer
+    :param rank: the default rank for ordering the document in queries:  if
+                 not passed, assigned a timestamp-based value by the server.
+                 See the ``rank`` entry on the page above for details.
+    """
+    def __init__(self, name, index, rank=None):
+        self.name = name
+        self.index = index
+        self.rank = rank
+        self.fields = {}
+
+    @classmethod
+    def from_api_repr(cls, resource, index):
+        """Factory:  construct a document given its API representation
+
+        :type resource: dict
+        :param resource: document resource representation returned from the API
+
+        :type index: :class:`gcloud.search.index.Index`
+        :param index: Index holding the document.
+
+        :rtype: :class:`gcloud.search.document.Document`
+        :returns: Document parsed from ``resource``.
+        """
+        name = resource.get('docId')
+        if name is None:
+            raise KeyError(
```

Comment:
If I were starting fresh, I wouldn't bother adding the custom error message:  the traceback supplies the same information if we just use `resource['docId']`.  Trying to be consistent across the APIs matters too, though.